### ISSUE TITLE
Extending vips for multiple images

### ISSUE THREAD

**Issue by prasadbandodkar** [libvips/libvips#4078]
Hi,

The Issues section has been very helpful, but I couldn't solve this problem. I am trying to understand how I should extend vips by sending multiple images. After following the tutorial here - https://www.libvips.org/API/current/extending.html - this is what I came up with for a sample function: vips_subtract.

`
static int sub_generate(VipsRegion *out_region, void *vseq, void *a, void *b, gboolean *stop)
{
     /* The area of the output region we have been asked to make.
    */
    VipsRect *r = &out_region->valid;

    /* The sequence value ... the thing returned by vips_start_one().
    */
    VipsRegion **ir = (VipsRegion **) vseq;

    VipsImage *in = (VipsImage *) a;
    int line_size = r->width * in->Bands;

    int x, y;

    /* Request matching part of input region.
    */
    if (vips_reorder_prepare_many(out_region->im, ir, r))
        return -1;

    for (y = 0; y < r->height; y++) 
    {
        // print y
        std::cout << "y: " << y << std::endl;
        unsigned char* p = (unsigned char *) VIPS_REGION_ADDR(ir[0], r->left, r->top + y);
        unsigned char* s = (unsigned char *) VIPS_REGION_ADDR(ir[1], r->left, r->top + y);
        unsigned char* q = (unsigned char *) VIPS_REGION_ADDR(out_region, r->left, r->top + y);

        for (x = 0; x < line_size; x++)
            q[x] = s[x] - p[x];
    }

    return 0;
}
`
This is not working. What am I doing wrong?

**Comment by jcupitt** [comment]
Hello @prasadbandodkar, could you explain what you are trying to achieve? I'm not sure I understand.

You'll need to post a complete program I can compile before I can test it.

**Comment by prasadbandodkar** [comment]
Hi @jcupitt, 

I am basically trying to understand how I can access individual pixel data from multiple input images and write to a single output image without creating a GObject. Therefore, I am trying to add or subtract two images by accessing individual pixel data instead of using functions like vips_add. I used the add.c and arithemetic.c to write it. However, since add.c depends on the arithmetic object to run, I didn't quite understand how the vips_add works if I were to implement it as a standalone program. I was able to run a program to ape vips_invert as per the tutorial - https://www.libvips.org/API/current/extending.html.

Here is the gist: https://gist.github.com/prasadbandodkar/b1d796f4a226204413eaef2e21ad978b

**Comment by jcupitt** [comment]
Ah OK. Why don't you want to make a gobject? It would be the simplest way to get what you want.

**Comment by prasadbandodkar** [comment]
I am using OpenCV in C++ to perform some image analysis operations for small images. I want to extend it to large images using vips for image handling. I want to access the individual pixels so I can use the generate function as the smallest unit of the image that I can call OpenCV methods on. Creating a GObject here would make things complicated I think. What do you think is the best way to execute this?

**Comment by jcupitt** [comment]
It would depend on what opencv operations you want to use. 

To use libvips generate, you need to be able to write a function like this:

```
# generate a rect of pixels for this image
def my_generate_function(image, rect):
    # what rect do we need from the input image to make this part of the output?
    input_rect = find_input_rect(rect)

    # ask libvips to make those pixels for us
    input_pixels = image.input.compute(input_rect)

    # do our computation
    output_pixels = do_my_computation(rect, input_pixels)

    return output_pixels
```

In other words, your opencv operation has to be able to generate any subregion of the output given only a small part of the input. Many useful opencv operations are NOT like this, unfortunately, and won't work in this style.

**Comment by prasadbandodkar** [comment]
In my case, rect and input_rect are the same. I need to work on the same rectangluar area of pixels in both input and output images. I think as long as I can write a vips_add kind of a standalone function, I should be able to make it work with my opencv methods. Could you please look at the gist I shared and point out what it is I am doing wrong? (I just discovered that in line 92 the array should be NULL terminated)

**Comment by jcupitt** [comment]
Are you certain that each output pixel depends on exactly one input pixel and nothing else? If that's the case, it must be a very simple operation, and you could almost certainly code it up yourself in one or two lines of C++. Can you give a specific example?

I'm out of time today, but I'll look at your gist tomorrow.

**Comment by prasadbandodkar** [comment]
No, but i intend to take all pixels in 'rect' of the input, convert it into a cv::Mat, process it in OpenCV, convert it back to a VipsRegion and write the pixels to the 'rect' region of the output. Will this not work?

**Comment by jcupitt** [comment]
Probably not -- you'll see a lot of terrible artifacts along the edges where the tiles join.

**Comment by jcupitt** [comment]
(though how bad they are depends on the specific operation, you'd need to give an example)

**Comment by jcupitt** [comment]
I reworked your gist and added it to the libvips `examples/` area, I hope that's OK:

https://github.com/libvips/libvips/blob/master/examples/my-add.c

I made a few changes:

- I made it only output the pixels in both inputs, so you can skip the `crop`
- you must create the output image before you call `vips_image_pipeline_array()`
- and after making the image and setting the fields to their default values with pipeline, you need to override any fields that will change (the format in your case)
- and a few small cleanups to make it prettier
